From 297062071ccbfafb4dfcc7438a87606333fa76b5 Mon Sep 17 00:00:00 2001
From: Andrew Kelley <andrew@ziglang.org>
Date: Mon, 8 Sep 2025 18:44:50 -0700
Subject: [PATCH 109/133] Sema: avoid ptr_add/ptr_sub instructions void elem
 type
X-Developer-Signature: v=1; a=openpgp-sha256; l=3369; i=xtex@astrafall.org;
 h=from:subject; bh=G8OegNA7/7R+iB/z6RUUoGP0DbNwI5/XmXRhHvElqJI=;
 b=owGbwMvMwCW2U4Ij7wZL9ETG02pJDJn78nhWX1n4Wv59/I4lJ5quWZ64uOel+bya5faLTqg5P
 VyQeDQ2tmMiC4MYF4OlmCJLkWGDN6tOOr/osnJZmDmsTCBDpEUaGBgYGFgY+HIT80qNdIz0TLUN
 9QwNdYx0jBi4OAVgqvd3MDI888/1k3Q6ccG7ficfr/K5jWmbL4jfk3E5f132Tck6Y3sLRoaPqkH
 C+6JueaqE13XKdpy6VX7Bb8Plp682zP++uuTH0T28AA==
X-Developer-Key: i=xtex@astrafall.org; a=openpgp;
 fpr=7231804B052C670F15A6771DB918086ED8045B91

(cherry picked from commit c0bbddb0071830eb9a6afd96124a24176f263b6a)
---
 src/Air.zig  | 30 ++++++++++++++++++------------
 src/Sema.zig |  6 +++++-
 2 files changed, 23 insertions(+), 13 deletions(-)

diff --git a/src/Air.zig b/src/Air.zig
index 77080386384d..72c73cc7613e 100644
--- a/src/Air.zig
+++ b/src/Air.zig
@@ -166,19 +166,25 @@ pub const Inst = struct {
         mod,
         /// Same as `mod` with optimized float mode.
         mod_optimized,
-        /// Add an offset to a pointer, returning a new pointer.
-        /// The offset is in element type units, not bytes.
-        /// Wrapping is illegal behavior.
-        /// The lhs is the pointer, rhs is the offset. Result type is the same as lhs.
-        /// The pointer may be a slice.
-        /// Uses the `ty_pl` field. Payload is `Bin`.
+        /// Add an offset, in element type units, to a pointer, returning a new
+        /// pointer. Element type may not be zero bits.
+        ///
+        /// Wrapping is illegal behavior. If the newly computed address is
+        /// outside the provenance of the operand, the result is undefined.
+        ///
+        /// Uses the `ty_pl` field. Payload is `Bin`. The lhs is the pointer,
+        /// rhs is the offset. Result type is the same as lhs. The operand may
+        /// be a slice.
         ptr_add,
-        /// Subtract an offset from a pointer, returning a new pointer.
-        /// The offset is in element type units, not bytes.
-        /// Wrapping is illegal behavior.
-        /// The lhs is the pointer, rhs is the offset. Result type is the same as lhs.
-        /// The pointer may be a slice.
-        /// Uses the `ty_pl` field. Payload is `Bin`.
+        /// Subtract an offset, in element type units, from a pointer,
+        /// returning a new pointer. Element type may not be zero bits.
+        ///
+        /// Wrapping is illegal behavior. If the newly computed address is
+        /// outside the provenance of the operand, the result is undefined.
+        ///
+        /// Uses the `ty_pl` field. Payload is `Bin`. The lhs is the pointer,
+        /// rhs is the offset. Result type is the same as lhs. The operand may
+        /// be a slice.
         ptr_sub,
         /// Given two operands which can be floats, integers, or vectors, returns the
         /// greater of the operands. For vectors it operates element-wise.
diff --git a/src/Sema.zig b/src/Sema.zig
index d50770d6b8d0..d0d8c9b26b3f 100644
--- a/src/Sema.zig
+++ b/src/Sema.zig
@@ -16046,7 +16046,6 @@ fn splat(sema: *Sema, ty: Type, val: Value) !Value {
 fn analyzeArithmetic(
     sema: *Sema,
     block: *Block,
-    /// TODO performance investigation: make this comptime?
     zir_tag: Zir.Inst.Tag,
     lhs: Air.Inst.Ref,
     rhs: Air.Inst.Ref,
@@ -16245,6 +16244,11 @@ fn analyzePtrArithmetic(
     const ptr_info = ptr_ty.ptrInfo(zcu);
     assert(ptr_info.flags.size == .many or ptr_info.flags.size == .c);
 
+    if ((try sema.typeHasOnePossibleValue(.fromInterned(ptr_info.child))) != null) {
+        // Offset will be multiplied by zero, so result is the same as the base pointer.
+        return ptr;
+    }
+
     const new_ptr_ty = t: {
         // Calculate the new pointer alignment.
         // This code is duplicated in `Type.elemPtrType`.
-- 
2.52.0

